File: 199-binary-tree-right-side-view/binary-tree-right-side-view.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public void soln(TreeNode root,List<Integer> ans,int rightMaxDepth[],int curDepth){
        if(root==null) return ;
        if(curDepth>rightMaxDepth[0]){
            ans.add(root.val);
            rightMaxDepth[0]=curDepth;
        }
        soln(root.right,ans,rightMaxDepth,curDepth+1);
        soln(root.left,ans,rightMaxDepth,curDepth+1);
    }
    public List<Integer> rightSideView(TreeNode root) {
        List<Integer> ans = new ArrayList<>();
        int rightMaxDepth[] = {-1};
        soln(root,ans,rightMaxDepth,0);
        return ans;
    }
}

File: 199-binary-tree-right-side-view/binary-tree-right-side-view.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    void soln(TreeNode* root,vector<int> &ans, int &maxD,int currD){
        if(root==nullptr) return;
        if(currD>maxD){
            ans.push_back(root->val);
            maxD = currD;
        }
        soln(root->right,ans,maxD,currD+1);
        soln(root->left,ans,maxD,currD+1);
    }
    vector<int> rightSideView(TreeNode* root) {
        vector<int> ans;
        int maxD = 0;
        soln(root,ans,maxD,1);
        return ans;
    }
};